js/web/lib/wasm/jsep/backend-webgpu.ts

```diff
@@ -94,11 +94,27 @@ const getProgramInfoUniqueKey =
       return key;
     };
 
+export class AdapterInfo {
```

Comment:
why not just use the webgpu's adapter info object and type? I didn't see benefit using a class wrapper.

Reply:
I see. In general I think it is OK to use a set of helper functions for this purpose. However I have a few concerns:
- I prefer to split definitions and implementation.
  - Using an interface helps to keep types.ts including types only (which compiles to nothing in JS)
- if this type is used to control feature ON/OFF, if may be a good idea to expose specific functions like `enableVectorizedGroupedConv(): Promise<boolean>`

Reply:
the interface can be something like `OptimizationFeatureManager` and multiple different implementation like `AdaptorInfoBasedFeatureManager` or `TestProgramBasedFeatureManager` can implement the interface, based on user's choice.

Reply:
I'm not sure if we need to expose this kind of complexity to developers. At least this is not my intention at this stage. I should be more specific, but helper functions just mean some easy-to-use functions based on GPUAdapterInfo. For example, I introduce isVendor() here to help the kernel to check the vendor easier. Later maybe we need another function to return the GPU generations, so that some optimizations can choose different paths based on different GPU gens.
I don't see the necessity now to expose kernel specific controls at API level. In stead, I just want to pass enough info, like the AdapterInfo here, to kernels, so that they may do whatever they want, while we keep all the logics within specific kernels' implementations.